provider/core/src/varule_traits.rs

```diff
@@ -0,0 +1,121 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use zerovec::ule::VarULE;
+
+#[cfg(feature = "alloc")]
+use zerovec::{maps::ZeroMapKV, ZeroMap2d};
+
+/// A trait that associates a [`VarULE`] type with a data struct.
+///
+/// Some data structs can be represented compactly as a single [`VarULE`],
+/// such as `str` or a packed pattern. This trait allows for data providers
+/// to use storage optimizations for such types.
```

Comment:
issue: you frame this as "storage optimizations", but that's not why we're doing this.

Reply:
no, this is a compile time improvement, at the expense of runtime performance

I have not seen any data that this reduces storage size

Reply:
It is a storage optimization that reduces compile times and binary size.

Reply:
I removed the word "storage". The point of this remark is that it can be used for optimizations at the discretion of the data provider.

Reply:
As we discussed in the WG meeting, I previously shared what I consider sufficient preliminary data in the main issue #5230, and more data can be gathered once this lands and is integrated into places such as datetime.